// @larix/zen/lib/plugins/AngularPlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var fs = require("fs");
var JSRuleFinder_1 = require("./shared/JSRuleFinder");
var AngularPlugin = (function () {
    function AngularPlugin() {
    }
    AngularPlugin.prototype.configure = function (builder, zen) {
        var stack = builder.stack;
        if (stack.hasAll(['angular', 'webpack'])) {
            var webpack = builder.require('webpack');
            var jsRuleFinder = new JSRuleFinder_1.default(builder);
            var tsRule = jsRuleFinder.findAndCreateTSRule();
            builder.config = zen.merge(builder.config, {
                module: {
                    rules: [
                        {
                            test: tsRule.test,
                            use: { loader: 'angular2-template-loader', options: zen.createConfig(builder, 'angular2Template', {}) }
                        },
                        {
                            test: /[\\\/]@angular[\\\/]core[\\\/].+\.js$/,
                            parser: { system: true }
                        }
                    ]
                },
                plugins: [
                    new webpack.ContextReplacementPlugin(/angular[\\\/]core/)
                ]
            });
            if (!stack.hasAny('dll') && stack.hasAny('web')) {
                var VirtualModules = builder.require('webpack-virtual-modules');
                var polyfillCode = fs.readFileSync(require.resolve('./angular/angular-polyfill.js')).toString();
                builder.config = zen.merge(builder.config, {
                    module: {
                        rules: [
                            {
                                test: /\.html$/,
                                loader: 'html-loader',
                                options: zen.createConfig(builder, 'html', {})
                            }
                        ]
                    },
                    plugins: [new VirtualModules({ 'node_modules/@virtual/angular-polyfill.js': polyfillCode })]
                });
                builder.config = zen.merge({
                    entry: {
                        index: ['@virtual/angular-polyfill']
                    }
                }, builder.config);
            }
        }
    };
    return AngularPlugin;
}());
exports.default = AngularPlugin;
