// @larix/zen/lib/plugins/ReactHotLoaderPlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var JSRuleFinder_1 = require("./shared/JSRuleFinder");
var ReactHotLoaderPlugin = (function () {
    function ReactHotLoaderPlugin() {
    }
    ReactHotLoaderPlugin.prototype.configure = function (builder, zen) {
        var stack = builder.stack;
        if (stack.hasAll(['react-hot-loader', 'webpack']) && zen.dev && !zen.test && !stack.hasAny('dll')) {
            builder.config = zen.mergeWithStrategy({
                entry: 'prepend'
            }, builder.config, {
                entry: {
                    index: ['react-hot-loader/patch']
                }
            });
            var jsRuleFinder = new JSRuleFinder_1.default(builder);
            var jsRule = jsRuleFinder.findAndCreateJSRule();
            var isBabelUsed = jsRule.use.loader && jsRule.use.loader.indexOf('babel') >= 0;
            jsRule.use = zen.merge(jsRule.use, {
                options: {
                    plugins: [isBabelUsed ? 'react-hot-loader/babel' : 'react-hot-loader/webpack']
                }
            });
        }
    };
    return ReactHotLoaderPlugin;
}());
exports.default = ReactHotLoaderPlugin;
