// @larix/zen/lib/plugins/VuePlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var VuePlugin = (function () {
    function VuePlugin() {
    }
    VuePlugin.prototype.configure = function (builder, zen) {
        var stack = builder.stack;
        if (stack.hasAll(['vue', 'webpack'])) {
            var webpack = builder.require('webpack');
            var VueLoaderPlugin = builder.require('vue-loader/lib/plugin');
            builder.config = zen.merge(builder.config, {
                module: {
                    rules: [
                        {
                            test: /\.vue$/,
                            use: { loader: 'vue-loader', options: zen.createConfig(builder, 'vue', {}) }
                        }
                    ]
                },
                resolve: {
                    alias: {
                        vue$: 'vue/dist/vue.esm.js'
                    }
                },
                plugins: [new VueLoaderPlugin()]
            });
        }
    };
    return VuePlugin;
}());
exports.default = VuePlugin;
