// @larix/zen/lib/Zen.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var merge = require("webpack-merge");
var Zen = (function () {
    function Zen(cwd, cmd) {
        this.cmd = cmd;
        this.cwd = cwd;
        this.dev = ['watch', 'start', 'test'].indexOf(this.cmd) >= 0;
        this.test = this.cmd === 'test';
        this.watch = ['watch', 'start'].indexOf(this.cmd) >= 0;
    }
    Zen.prototype.createConfig = function (builder, tool, config) {
        var _a = builder[tool + 'Config'] || { merge: {} }, mergeStrategy = _a.merge, configOverrides = __rest(_a, ["merge"]);
        return this.mergeWithStrategy(mergeStrategy, config, configOverrides);
    };
    Zen.prototype.merge = function (config, overrides) {
        return merge.smart(config, overrides);
    };
    Zen.prototype.mergeWithInStrategy = function (config, overrides) {
        var _a = overrides || { merge: {} }, mergeStrategy = _a.merge, configOverrides = __rest(_a, ["merge"]);
        return this.mergeWithStrategy(mergeStrategy, config, configOverrides);
    };
    Zen.prototype.mergeWithStrategy = function (strategy, config, overrides) {
        return merge.smartStrategy(strategy)(config, overrides);
    };
    return Zen;
}());
exports.default = Zen;
