// @larix/zen/lib/plugins/shared/JSRuleFinder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var resolveModule_1 = require("./resolveModule");
var JSRuleFinder = (function () {
    function JSRuleFinder(builder) {
        this.builder = builder;
    }
    JSRuleFinder.prototype.findJSRule = function () {
        if (!this.jsRule) {
            var jsCandidates = ['\\.js$/', '\\.jsx?$/'];
            for (var _i = 0, _a = this.builder.config.module.rules; _i < _a.length; _i++) {
                var rule = _a[_i];
                for (var _b = 0, jsCandidates_1 = jsCandidates; _b < jsCandidates_1.length; _b++) {
                    var candidate = jsCandidates_1[_b];
                    if (String(rule.test).indexOf(candidate) >= 0) {
                        this.jsRule = rule;
                        break;
                    }
                }
            }
        }
        return this.jsRule;
    };
    JSRuleFinder.prototype.createJSRule = function () {
        var _this = this;
        if (this.jsRule) {
            throw new Error('js rule already exists!');
        }
        this.jsRule = {
            test: /\.js$/,
            exclude: function (modulePath) { return resolveModule_1.default(_this.builder, modulePath).moduleType === resolveModule_1.ModuleType.NormalNodeModule; }
        };
        this.builder.config.module.rules = this.builder.config.module.rules.concat(this.jsRule);
        return this.jsRule;
    };
    JSRuleFinder.prototype.findAndCreateJSRule = function () {
        return this.findJSRule() || this.createJSRule();
    };
    JSRuleFinder.prototype.findTSRule = function () {
        if (!this.tsRule) {
            var jsCandidates = ['\\.ts$/', '\\.tsx?$/'];
            for (var _i = 0, _a = this.builder.config.module.rules; _i < _a.length; _i++) {
                var rule = _a[_i];
                for (var _b = 0, jsCandidates_2 = jsCandidates; _b < jsCandidates_2.length; _b++) {
                    var candidate = jsCandidates_2[_b];
                    if (String(rule.test).indexOf(candidate) >= 0) {
                        this.tsRule = rule;
                        break;
                    }
                }
            }
        }
        return this.tsRule;
    };
    JSRuleFinder.prototype.createTSRule = function () {
        var _this = this;
        if (this.tsRule) {
            throw new Error('ts rule already exists!');
        }
        this.tsRule = {
            test: /\.ts$/,
            exclude: function (modulePath) {
                var moduleType = resolveModule_1.default(_this.builder, modulePath).moduleType;
                return moduleType === resolveModule_1.ModuleType.NormalNodeModule || moduleType === resolveModule_1.ModuleType.TranspiledNodeModule;
            }
        };
        this.builder.config.module.rules = this.builder.config.module.rules.concat(this.tsRule);
        return this.tsRule;
    };
    JSRuleFinder.prototype.findAndCreateTSRule = function () {
        return this.findTSRule() || this.createTSRule();
    };
    Object.defineProperty(JSRuleFinder.prototype, "extensions", {
        get: function () {
            var result = [];
            this.findJSRule();
            this.findTSRule();
            var jsTestStr = String(this.jsRule ? this.jsRule.test : 'js');
            var tsTestStr = String(this.tsRule ? this.tsRule.test : '');
            if (tsTestStr.indexOf('tsx') >= 0) {
                result.push('tsx');
            }
            if (jsTestStr.indexOf('jsx') >= 0) {
                result.push('jsx');
            }
            if (tsTestStr.indexOf('ts') >= 0) {
                result.push('ts');
            }
            if (jsTestStr.indexOf('js') >= 0) {
                result.push('mjs');
                result.push('js');
            }
            return result;
        },
        enumerable: true,
        configurable: true
    });
    return JSRuleFinder;
}());
exports.default = JSRuleFinder;
